File: lab6-openmp/lab6-openmp_c/omp_bug2.c
/* FILE: omp_bug2.c
   DESCRIPTION:
      Another OpenMP program with a bug. 
   AUTHOR: Blaise Barney 
   UPDATED: Daniel R. Reynolds, 1/13/2013 */

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

int main (int argc, char *argv[]) {

  /* local variables */
  int nthreads, i, tid;
  float total;

  /* Spawn parallel region */
# pragma omp parallel 
  {
    /* Obtain thread number */
    tid = omp_get_thread_num();

    /* Only master thread does this */
    if (tid == 0) {
      nthreads = omp_get_num_threads();
      printf("Number of threads = %d\n", nthreads);
    }

    /* Everyone does this */
    printf("Thread %d is starting...\n",tid);

#   pragma omp barrier

    /* do some work */
    total = 0.0;
#   pragma omp for schedule(dynamic,10)
    for (i=0; i<1000000; i++)  total = total + i*1.0;
    
    printf ("Thread %d is done! Total= %e\n",tid,total);

  } /* End of parallel region */

  return 0;
}  /* end main */

File: lab6-openmp/lab6-openmp_f90/omp_bug2.f90
! -*- Mode: Fortran90; -*-
!=================================================================

program BUG2
  !-----------------------------------------------------------------
  ! FILE: omp_bug2.f90
  ! DESCRIPTION:
  !   Another OpenMP program with a bug
  ! AUTHOR: Blaise Barney  1/7/04
  ! Upgraded to F90, free-source form: Daniel R. Reynolds, 2/2009
  !-----------------------------------------------------------------
  !======= Inclusions ===========

  !======= Declarations =========
  implicit none

  integer :: NThreads, i, TID
  real*8 :: total
  integer, external :: omp_get_num_threads, omp_get_thread_num

  !======= Internals ============

  ! Spawn parallel region
  !$omp parallel

  ! Obtain thread number
  TID = omp_get_thread_num()

  ! Only master thread does this
  if (TID == 0) then
     NThreads = omp_get_num_threads()
     print *, 'Number of threads = ', NThreads
  end if
  print *, 'Thread ',TID,'is starting...'

  !$omp barrier

  ! Do some work
  total = 0.0
  !$omp do schedule(dynamic,10)
  do i=1,1000000
     total = total + i*1.d0
  end do

  write(*,*) 'Thread',TID,' is done! Total= ',total
  !$omp end parallel

end program BUG2
!=================================================================
